File: 20131209/AM/131209_tests/subway/check/subway2.java
import java.util.*;
import java.io.*;

public class subway {
  public static String canonical(String s) {
    Vector<String> parts = new Vector<String>();
    int zeros = 0;
    int ones = 0;
    int current_start = 1;

    for (int i = 0; i < s.length(); ++i) {
      if (s.charAt(i) == '0') {
        zeros += 1;
      } else {
        ones += 1;
      }

      if (zeros == ones) {
        parts.add("0" + canonical(s.substring(current_start, i)) + "1");
        current_start = i + 2;
      }
    }

    String ans = "";
    Collections.sort(parts);

    for (int i = 0; i < parts.size(); ++i) {
      ans += parts.get(i);
    }

    return ans;
  }

  public static void main(String[] args) {
    Scanner fin = new Scanner(System.in);
    int n = fin.nextInt();

    Set<String> unique = new TreeSet<String>();

    for (int i = 0; i < n; ++i) {
      String s = fin.next();

      unique.add(canonical(s));
    }

    System.out.println(unique.size());
  }
}

File: 20131209/AM/131209_tests/subway/check/subway2.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <string>
#include <vector>

using namespace std;

string canonical(const string& s) {
  vector<string> parts;
  int zeros = 0;
  int ones = 0;
  int current_start = 1;

  for (size_t i = 0; i < s.size(); ++i) {
    if (s[i] == '0') {
      zeros += 1;
    } else {
      ones += 1;
    }

    if (zeros == ones) {
      parts.push_back("0" + canonical(s.substr(current_start, i-current_start+1)) + "1");
      current_start = i + 2;
    }
  }

  string ans;
  sort(parts.begin(), parts.end());

  for (size_t i = 0; i < parts.size(); ++i) {
    ans += parts[i];
  }

  return ans;
}

int main(int argc, char** argv) {
  int n;
  cin >> n;

  set<string> unique;

  for (int i = 0; i < n; ++i) {
    string s;
    cin >> s;

    unique.insert(canonical(s));
  }

  cout << unique.size() << endl;
}
